Fix a pause in the event loop when clicking the title bar on windows

aloucks opened the pull request:
When clicking the title bar on Windows, to drag the window, there is a noticeable ~500ms pause in continuous redraw requests. This was fixed in #839 and then regressed in #1852. The cursor blinks in both cases and is unrelated. The regression made the blink happen after the pause instead of immediately.

- [x] Tested on all platforms changed
- [x] Added an entry to the `changelog` module if knowledge of this change could be valuable to users

madsmtm commented:
Will this fix https://github.com/rust-windowing/winit/issues/1885 too?

aloucks commented:
> Will this fix #1885 too?

Yes

madsmtm reviewed (commented):
So, I would quite like to get this fix in, as I believe it's desired by Bevy people, but we don't really have a Windows maintainer at the moment, and I don't feel like I understand enough of what's going on here myself to confidently merge this.

Maybe you could expand on why the `PostMessageW` is necessary in the first place (why is the default handler not sufficient?), why passing `lparam = 0` is correct, and if you know, how we'd go about fixing the blinking then? Ideally as code-comments with links to official docs, makes it easier to track in the future.

Also CC author of https://github.com/rust-windowing/winit/pull/1852 @Ssaely

aloucks commented:
> So, I would quite like to get this fix in, as I believe it's desired by Bevy people, but we don't really have a Windows maintainer at the moment, and I don't feel like I understand enough of what's going on here myself to confidently merge this.
> 
> Maybe you could expand on why the `PostMessageW` is necessary in the first place (why is the default handler not sufficient?), why passing `lparam = 0` is correct, and if you know, how we'd go about fixing the blinking then? Ideally as code-comments with links to official docs, makes it easier to track in the future.
> 
> Also CC author of #1852 @Ssaely

To be honest, I [fixed this 6 years ago](https://github.com/rust-windowing/winit/pull/839) and I don't really remember the details. I noticed it had regressed while exploring something else. The blinking happens even when the entire `WM_NCLBUTTONDOWN` event handler is commented out. The primary difference is that with the change, it blinks immediately instead of after the short pause.

The short pause is ingrained in how Windows works. If you start dragging the window immediately when you click, you won't notice a pause. I think that's why posting a dummy `WM_MOUSEMOVE` message eliminates the delay. The `lparam` value of `WM_MOUSEMOVE` is the `(x, y)` coordinates of the **client area**, where as the `WM_NCLBUTTONDOWN` `lparam` is the `(x, y)` in **screen coordinates**.

I *think* what's happening is that the dummy `WM_MOUSEMOVE` simulates a window drag, which cancels the pause, but because you're in the non-client-area, the message is never actually put onto the event queue so you never see the dummy event. I'm guessing that windows computing a delta to determine drag status and `(0, 0)` is always going to be far enough away that the simulated bogus drag cancels the pause.

In summary, this is absolutely a hacky work-around :)

**The actual robust and generally accepted solution is to execute your simulation and rendering logic from a different thread.**

madsmtm commented:
Thanks a lot for the context! I'd be okay with merging this if you included basically the block of text you wrote here as a code comment.

aloucks commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -39,24 +39,7 @@ use windows_sys::Win32::UI::Input::Touch::{
 };
 use windows_sys::Win32::UI::Input::{RAWINPUT, RIM_TYPEKEYBOARD, RIM_TYPEMOUSE};
 use windows_sys::Win32::UI::WindowsAndMessaging::{
```
I'm not sure what happened here. The requirement to use nightly rustfmt is kind of frustrating,

aloucks commented:
@madsmtm 
I added additional comments and updated the window example to better detect when the event loop freezes. There's now an option to animate the fill color.

aloucks commented:
> Could I get you to you update the `application.rs` example instead?

I was actually going to ask about adding back more of the single purpose examples. That applicaiton.rs example is overwhelming. I can see how it may be useful in some cases, but the individual examples are sorely missed.

kchibisov commented:
one line examples should be in a `method` docs respectively, no point to create too many files to show 1 function call.

aloucks commented:
> no point to create too many files to show 1 function call.

Respectfully, I disagree with that philosophy when it comes to examples. Short concise and complete examples are easier for someone unfamiliar with the code to understand and help communicate what is relevant to the the goal of the example. They also serve as a way to test functionality independently. I'm a big fan of examples in documentation itself, but they are generally not complete or aren't helpful for iterating on a change. I see value in a larger integrated example; I just don't agree with the idea that there shouldn't be smaller examples too.

In any case, I've added new actions in `application.rs` for toggling continuous redraw as well as toggling the animated fill color.

filnet commented on src/platform_impl/windows/event_loop.rs:
```diff
@@ -1135,6 +1135,31 @@ unsafe fn public_window_callback_inner(
+                //
+                // When right-click the title bar, the system window menu is presented to the user,
+                // and the modal event loop begins. This dummy event does *not* prevent the freeze
+                // in the main event loop imposed caused by that popup menu.
```
imposed or caused ?

aloucks replied:
Updated

madsmtm commented:
> > no point to create too many files to show 1 function call.
> 
> Respectfully, I disagree with that philosophy when it comes to examples.

I tend to agree, but this is the wrong place/time to discuss that (we can do that in another issue/PR if you want).

> In any case, I've added new actions in application.rs for toggling continuous redraw as well as toggling the animated fill color.

Thanks! Could I get you to remove it from `examples/window.rs`, then? I'd like to have at least one example that is as simple as possible (regardless of the number of examples we should have).

madsmtm commented:
> Thanks! Could I get you to remove it from `examples/window.rs`, then? I'd like to have at least one example that is as simple as possible (regardless of the number of examples we should have).

Nvmd, I did it myself, hope you don't mind.

Again, happy to discuss changing our examples, but should be done in a separate issue/PR.
